pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,463 @@
+		RequestCannotYetBeFulfilled,
+		OnlyRequesterCanIncreaseFee,
+		RequestHasNotExpired,
+		RequestExecutionOOG,
+		InstantRandomnessNotAvailable,
+		RandomnessResultDNE,
+		RandomnessResultNotFilled,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RandomnessRequestedCurrentBlock {
+			id: RequestId,
+			refund_address: H160,
+			contract_address: H160,
+			fee: BalanceOf<T>,
+			gas_limit: u64,
+			salt: H256,
+			earliest_block: T::BlockNumber,
+		},
+		RandomnessRequestedBabeOneEpochAgo {
+			id: RequestId,
+			refund_address: H160,
+			contract_address: H160,
+			fee: BalanceOf<T>,
+			gas_limit: u64,
+			salt: H256,
+			earliest_epoch: u64,
+		},
+		RandomnessRequestedBabeTwoEpochsAgo {
+			id: RequestId,
+			refund_address: H160,
+			contract_address: H160,
+			fee: BalanceOf<T>,
+			gas_limit: u64,
+			salt: H256,
+			earliest_epoch: u64,
+		},
+		RandomnessRequestedLocal {
+			id: RequestId,
+			refund_address: H160,
+			contract_address: H160,
+			fee: BalanceOf<T>,
+			gas_limit: u64,
+			salt: H256,
+			earliest_block: T::BlockNumber,
+		},
+		RequestFulfilled {
+			id: RequestId,
+		},
+		RequestFeeIncreased {
+			id: RequestId,
+			new_fee: BalanceOf<T>,
+		},
+		RequestExpirationExecuted {
+			id: RequestId,
+		},
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn requests)]
+	/// Randomness requests not yet fulfilled or purged
+	pub type Requests<T: Config> = StorageMap<_, Twox64Concat, RequestId, RequestState<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn request_count)]
+	/// Number of randomness requests made so far, used to generate the next request's uid
+	pub type RequestCount<T: Config> = StorageValue<_, RequestId, ValueQuery>;
+
+	/// Current local per-block VRF randomness
+	/// Set in `on_initialize`, before it will contain the randomness for this block
+	#[pallet::storage]
+	#[pallet::getter(fn local_vrf_output)]
+	pub type LocalVrfOutput<T: Config> = StorageValue<_, Option<T::Hash>, ValueQuery>;
+
+	/// Current VRF input
+	/// Set in `on_finalize` of last block
+	/// Used in `on_initialize` of current block to validate VRF output
+	#[pallet::storage]
+	#[pallet::getter(fn current_vrf_input)]
+	pub(crate) type CurrentVrfInput<T: Config> = StorageValue<_, VrfInput<Slot, T::Hash>>;
+
+	/// Relay time information to determine when to update randomness results based on when
+	/// epoch and relay block change
+	#[pallet::storage]
+	#[pallet::getter(fn relay_time)]
+	pub(crate) type RelayTime<T: Config> =
+		StorageValue<_, RelayTimeInfo<T::BlockNumber, u64>, ValueQuery>;
+
+	/// Ensures the mandatory inherent was included in the block
+	#[pallet::storage]
+	#[pallet::getter(fn inherent_included)]
+	pub(crate) type InherentIncluded<T: Config> = StorageValue<_, ()>;
+
+	/// Records whether this is the first block (genesis or runtime upgrade)
+	#[pallet::storage]
+	#[pallet::getter(fn not_first_block)]
+	pub type NotFirstBlock<T: Config> = StorageValue<_, ()>;
```

Comment:
I not have a strong opinion about this, I understand your idea, it allow to skip "initialization stuff" at the cost of an additional storage item. It's not very elegant, but it works.

Reply:
To be clear, the original purpose of the storage item was to not verify nor store the VrfOutput in the first block because there is no VrfInput set in the block before.

I removed the migration and genesis storage because it prevented the need for them as well. I will make a ticket for a follow up migration to remove the storage item and the extra if branch in `set_output` for example.